Add deprecation policy doc (#9723)
matrix-org/synapse: docs/deprecation_policy.md

Deprecation Policy for Platform Dependencies
============================================

> Synapse has a number of platform dependencies, including Python and PostgreSQL.
> This document outlines the policy towards which versions we support, and how we
> drop support for versions in the future.

Reviewer: Maybe?

```suggestion
Synapse has a number of platform dependencies, including Python and PostgreSQL.
This document outlines the policy towards which versions we support, and when we
drop support for versions in the future.
```

> This document outlines the policy towards which versions we support, and how we
> drop support for versions in the future.

Reviewer 2: ```suggestion
This document outlines the policy of which versions of specific dependencies
we support, and how we drop support for old versions in the future.
```

Policy
------

> For Python and PostgreSQL Synapse will follow the upstream support life cycles,
> i.e. when a version reaches End of Life Synapse will withdraw support for that
> version in future releases.

Reviewer: ```suggestion
Synapse follows the upstream support life cycles for Python and PostgreSQL,
i.e. when a version reaches End of Life Synapse will withdraw support for that
version in future releases.
```

Deprecation Policy for Platform Dependencies
============================================

Synapse has a number of platform dependencies, including Python and PostgreSQL.
This document outlines the policy towards which versions we support, and how we
drop support for versions in the future.

Policy
------

For Python and PostgreSQL Synapse will follow the upstream support life cycles,
i.e. when a version reaches End of Life Synapse will withdraw support for that
version in future releases.

The team will not intentionally break compatibility for at least one month after
an EOL date, but does explicitly reserve the right to do so should security or

> other concerns arise in an unmaintained version of a dependency.

Reviewer 3: ...do we _actually_ want to box ourselves in to this, or should we drop when upstream drops?

Author: I'd me more than happy to drop this, it seems a bit pointless.

Reviewer 3: Make it so

> 

Reviewer 2: ```suggestion
```

Details on the upstream support life cycles for Python and PostgreSQL are
documented at https://endoflife.date/python and
https://endoflife.date/postgresql.

> 

Reviewer 2: Do we not want to comment on the policy for any other dependencies?

Reviewer 3: Do we have other deps we care about in this regard?

Reviewer 2: Probably not, but it may be nice to explicitly say that we don't.

Reviewer: We can always add more in the future. 🤷

Context
-------

It is important for system admins to have a clear understanding of the platform

> requirements of Synapse and deprecation policies so that they can effectively

Reviewer 2: ```suggestion
requirements of Synapse and its deprecation policies so that they can effectively
```
